packages/website/src/pages/playground.js

```diff
@@ -1,20 +1,51 @@
+import { deflate, inflate } from "pako";
+import { toByteArray, fromByteArray } from "base64-js";
 import React from "react";
 import Layout from "@theme/Layout";
 import { SquigglePlayground } from "../components/SquigglePlayground";
 
+function getHashData() {
+  if (typeof window === "undefined" || !window.location.hash) {
+    return {};
+  }
+  try {
+    const compressed = toByteArray(
+      decodeURIComponent(window.location.hash.slice(1))
+    );
+    const text = inflate(compressed, { to: "string" });
+    return JSON.parse(text);
+  } catch (err) {
+    console.error(err);
+    return {};
+  }
+}
+
+function setHashData(data) {
+  const text = JSON.stringify({ ...getHashData(), ...data });
+  const compressed = deflate(text, { level: 9 });
+  window.location.hash = encodeURIComponent(fromByteArray(compressed));
```

Comment:
This breaks the back button — it created a new entry on every edit.

It should be something like:
`window.location.replace( window.location.href.split('#')[0] + '#' + hash );`

But also, I'm not sure about changing the hash on every code change. Since hash is opaque, it's not clear what exactly will be shared, and it's easy to share your model by accident if you just want to show the playground UI to someone. It also fills my browser history with nonsense urls, but that might go away after the back button fix.

Maybe an icon that would copy an url with hash on click would be better?

Reply:
> This breaks the back button — it created a new entry on every edit.

Good catch! I thought changing just the hash wouldn't mess with the history, since it doesn't trigger a page reload, definitely should be fixed

> I'm not sure about changing the hash on every code change ... it's easy to share your model by accident


For prior art, the [Typescript playground](https://www.typescriptlang.org/play?#code/PTAEAkFEBloeVAZQIoFUCSBxT1IEIAoAkUAFQE8AHAU0QGMAnAS0oBdQmBnUAQyVYYBXOq0EMeAGzJVqScp1bUAtgDpQAQVAKhIsZNCsZxMJ3mKloJdR4A7bqwAWPdgHcH1G6DoB7JZR7MNgDmBjKcADTSNPTMbKDeNhLkxgY8ANbU3Ew2rN68dD6COQbulsoARtQM3AklsoY0KkQkpA5cHFmePjniCgZ5Nr7Z+g2ypgrKEaBuVbLk3oIphRIAJl4M1ooGLnmj3OWC7Mtrg+w8nJxMQZ6Oyv2g1nQOKd63DGqI1LLUAB48fhJqAAuQZKYYSAC0DWyQWaYAAYt4GA8-gDqJFbpx6jsODkqgAzHh0TK8DZeXyUQGKJIpAS2Tj4qricqA3G8LQCYSicRSUZaMzKIFEbKKBiE4mgABCkikAG8CKBQCsmP9qKKgaAbIIlJUGABuAgAXwIIoJRNkiEo7jJ8sVytV6s12t1BuNBCpoHKMo10okUgAvKBZUqVVZHQBGAAMoENBo9nCtsw1lutskDwftYaqGoATNHY0QE6nQIGvX6DWWA1pExsDSl0PjprIeCs1tl2Xy3EwnhydNyZeRyTkeNluDL4vjaaUrDqqjVG76pLY1inZhj3J4mGcJJw8nRbJ7qCksexcof2ZX4sii7Mmia8WLzWRBJUgwqQw7s07Z-r34Dgo4Gpaj+rpEB6oiVBqpAvmmQYflmDAauGOaRP+QSAaAADMMZ1hBsGVhW46BnhdYkJK1D7oIWIXuOKzeCSpygE4ABuzagGhjjrh4HBHLYjGVCk5yXNc1BrGetwGDBoDMQEKostQajgN4LjUKxDCROO3iNhJM66pwLwLuOASyKOTArPUMEaTYay7jxXgHq+QlXDYol3i0MgxCw7DtD4fiycEDxEg4ZQ-myEl8jwQQjnYrApI8wWvNa9xqUw+KDrcTDItQACOgiSFu5BuWAmj4kUIhMLU7QHt45QAFYUd5ngAFI8DJnlxMudk+RSxkrCk7Z8JcYISAEoCEpwbQJGoADqW4JS5oBUdQpVSL8dIGMwdBpKSCzWVOsj+OISicEK7pqusmzUIuJagAAFJmapfsBuoAJQlgAfHdGaho9yKGi9cbnYwl2rmSgb3T9jrPVUkRLegNhPJkGrlN43iArYb3+p9tqgBs3KeN9n6IYtWLw4j3AAPzwb9oAAFSgJGKiYQAnKAGoPaKOFGnWwPOLQNawbziiLgaQtXURF186D1CkWAFDRIwXnTEwfq8H6yl3dDDBvWeOV5VIt1a5EKNo9YNgvf1NylIdRNTAc7CMYbzow56qPo+boAQp9Ts-hbKTy7QitxC4KtSMqnD7gwYmlCb7thaU+KZX0TnXFYOQpJUlHUVuADk3BqYOvlKLU+JIqALVtUHRzeOZ9ycGkLCgP4FwwikNvHdMG51OQOdkoxLmia5cKgIiWWopS9SlAE4g99wpcMJAQV5-ZfpeltTHnCUGwHQE-wRCkMkSII6K4uZPy8NZdRjYIqvTzwg4QhwjYB+17DKisNg57FJCcIIlCUEiN+XBI7KgCu3X6UxbieHmIIaYCxVgb1Yv0S2dAj611UlUQc3g2AVRuHkJQ6RJ6yFKgjVguDuAENYD2P+p0ADaYtFy3XDC9SIDCZS3RzC9AAuioeei8ni3VupWSIAB9bIZ9RGVk4JjbG75ui7kBCoCQ3gghCJlADI0GiUgADk8gJFkAPFY9gnCsCKqAAASmqMQNwwiknqBsPmawJBMAyPEOqDUpidVsIOZU+JGQbARpkQSZJfKHVEsrRwV9OCqjcfVEQDxcr5UMLjG+JIeAo1YneD0YtLErGuuDDmT1nbaw+l9amopWFoyRBqAARBsFYNSYxaLYarQMOTRIiwIO0vJEsWkSFlqAaaxYJKJ2qGcFuqcPCuCRGkbgt1biDhRpElishCnawzocKJFEEhrDopkT+7B5kx1ovRTgBzEFsR8Mo7WTQgA) does it when the code editor loses focus, the [rescript playground](https://rescript-lang.org/try) has a share button that also edits the URL.

I don't think the risk of accidentally sharing private data is very large, but adding a button can be a good affordance nevertheless.

I would personally go with the minimal version and improve on it after trying it out in practice, but also happy to add a button.

Reply:
Maybe we can prefix the hash with `#code=` to make it more transparent? (similar to what they do on the typescript playground)

Reply:
I agree about releasing the minimal version first. (And, to be clear, I'm not that bothered about editing the url automatically, just wanted to share my first impression).

Reply:
> when the code editor loses focus

That seems pretty fine to me, maybe we do it this way for now?